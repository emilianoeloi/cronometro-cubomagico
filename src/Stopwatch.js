import React, { Component } from 'react';
import { msToISOString } from './Common';

class Stopwatch extends Component {
  constructor(props) {
    super(props);
    this.start = this.start.bind(this);
    this.stop = this.stop.bind(this);
    this.reset = this.reset.bind(this);
    this.save = this.save.bind(this);
    this.resetTwoKeys = this.resetTwoKeys.bind(this);
    this.state = {secondsElapsed: 0};
    this.focused = true;

  }

  tick() {
    this.setState((prevState) => ({
      secondsElapsed: prevState.secondsElapsed + 1
    }));
  }

  reset() {
    this.setState({ secondsElapsed: 0});
  }

  start() {
    if (this.control.started) return;
    this.reset();
    this.interval = setInterval(() => this.tick(), 1000);
    this.control.started = true;
    document.querySelector('.App-logo').className += ' App-logo-anim';
  }
  save() {
    const {
      saveStopwatchTime,
    } = this.props;
    saveStopwatchTime(this.state.secondsElapsed);
  }

  stop() {
    clearInterval(this.interval);
    this.control.started = false
    document.querySelector('.App-logo').className = document.querySelector('.App-logo').className.replace(' App-logo-anim', '');
  }

  checkPCKeys(evt) {
    if (evt.ctrlKey) {
      if (!this.twoKeys.first) {
        this.twoKeys.first = true;
      } else if (this.twoKeys.first) {
        this.twoKeys.second = true;
      }
      if (this.twoKeys.first && this.twoKeys.second) {
        if (this.twoKeys.stop) {
          this.stop();
          this.twoKeys.stop = false;
        } else {
          this.start();
          this.twoKeys.stop = true;
        }
        this.twoKeys.first = false;
        this.twoKeys.second = false;
      }
    }
  }
  checkMacKeys(evt) {
    const key = evt.key;
    if (key === 'Meta') {
      if (!this.twoKeys.first) {
        this.twoKeys.first = true;
      } else if (this.twoKeys.first) {
        this.twoKeys.second = true;
      }
      if (this.twoKeys.first && this.twoKeys.second) {
        if (this.twoKeys.stop) {
          this.stop();
          this.twoKeys.stop = false;
        } else {
          this.start();
          this.twoKeys.stop = true;
        }
        this.twoKeys.first = false;
        this.twoKeys.second = false;
      }
    }
  }

  resetTwoKeys() {
    this.twoKeys = {
      first: false,
      second: false,
      stop: false,
    };
  }

  componentDidMount() {
    this.control = {
      started: false,
    }
    this.resetTwoKeys();
    document.onkeydown = (evt) =>  {
      if (!evt) evt = event;
      this.checkPCKeys(evt);
      this.checkMacKeys(evt);
    };

    window.onfocus = () => {
      this.focused = true;
    };
    window.onblur = () => {
      this.focused = false;
      console.info('window.onfocus');
      this.resetTwoKeys();
    };
  }

  render() {
    return (
      <div>
        <h2 id='stopwatch-display'>{msToISOString(this.state.secondsElapsed)}</h2>
        <button className="pure-button" id="stopwatch-start" onClick={this.start}>
          Iniciar
        </button> <button className="pure-button" id="stopwatch-stop" onClick={this.stop}>
          Parar
        </button> <button className="pure-button" id="stopwatch-saveTime" onClick={this.save}>
          Salvar Tempo
        </button> <button className="pure-button" id="stopwatch-discardTime" onClick={this.reset}>
          Descartar Tempo
        </button>
      </div>
    );
  }
}

export default Stopwatch;
