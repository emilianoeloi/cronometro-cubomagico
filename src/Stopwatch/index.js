import React, { Component } from 'react';
import ReactGA from 'react-ga';

import { msToISOString } from '../Common';

class Stopwatch extends Component {
  constructor(props) {
    super(props);
    this.start = this.start.bind(this);
    this.stop = this.stop.bind(this);
    this.reset = this.reset.bind(this);
    this.save = this.save.bind(this);
    this.resetTwoKeys = this.resetTwoKeys.bind(this);
    this.state = {secondsElapsed: 0};
    this.focused = true;
    this.step = 1000 / 16;
  }

  tick() {
    this.setState((prevState) => ({
      secondsElapsed: prevState.secondsElapsed + this.step
    }));
  }

  reset() {
    this.setState({ secondsElapsed: 0});
  }

  start() {
    if (this.control.started) return;
    this.reset();
    this.interval = setInterval(() => this.tick(), this.step);
    this.control.started = true;
    document.querySelector('.App-logo').className += ' App-logo-anim';
  }
  save() {
    const {
      saveStopwatchTime,
    } = this.props;
    saveStopwatchTime(this.state.secondsElapsed);
  }

  stop() {
    clearInterval(this.interval);
    this.control.started = false
    ReactGA.event({
      category: 'Stopwatch',
      action: 'stop',
      label: 'time',
      value: this.state.secondsElapsed
    });
    document.querySelector('.App-logo').className = document.querySelector('.App-logo').className.replace(' App-logo-anim', '');
  }

  checkPCKeys(evt) {
    if (evt.ctrlKey) {
      if (!this.twoKeys.first) {
        this.twoKeys.first = true;
      } else if (this.twoKeys.first) {
        this.twoKeys.second = true;
      }
      if (this.twoKeys.first && this.twoKeys.second) {
        if (this.twoKeys.stop) {
          this.stop();
          this.twoKeys.stop = false;
        } else {
          this.start();
          this.twoKeys.stop = true;
        }
        this.twoKeys.first = false;
        this.twoKeys.second = false;
      }
    }
  }
  checkMacKeys(evt) {
    const key = evt.key;
    if (key === 'Meta') {
      if (!this.twoKeys.first) {
        this.twoKeys.first = true;
      } else if (this.twoKeys.first) {
        this.twoKeys.second = true;
      }
      if (this.twoKeys.first && this.twoKeys.second) {
        if (this.twoKeys.stop) {
          this.stop();
          this.twoKeys.stop = false;
        } else {
          this.start();
          this.twoKeys.stop = true;
        }
        this.twoKeys.first = false;
        this.twoKeys.second = false;
      }
    }
  }

  resetTwoKeys() {
    this.twoKeys = {
      first: false,
      second: false,
      stop: false,
    };
  }

  componentDidMount() {
    this.control = {
      started: false,
    }
    this.resetTwoKeys();
    document.onkeydown = (evt) =>  {
      if (!evt) evt = event;
      this.checkPCKeys(evt);
      this.checkMacKeys(evt);
    };

    window.onfocus = () => {
      this.focused = true;
    };
    window.onblur = () => {
      this.focused = false;
      this.resetTwoKeys();
    };
  }

  render() {
    const {
      logged
    } = this.props;
    return (
      <div className="pure-g">
        <div className="pure-u-1">
          <h2 id='stopwatch-display'>{msToISOString(this.state.secondsElapsed)}</h2>
        </div>
        <div className="pure-u-1">
          <div className="l-box">
            <button className="pure-button button-start" id="stopwatch-start" onClick={this.start}>
              Iniciar
            </button> <button className="pure-button button-stop" id="stopwatch-stop" onClick={this.stop}>
              Parar
            </button>
          </div>
        </div>
        <div className="pure-u-1">
          <div className="l-box">
            <button disabled={!logged} className="pure-button pure-button-active" id="stopwatch-saveTime" onClick={this.save}>
              Salvar Tempo
            </button> <button className="pure-button pure-button-active" id="stopwatch-discardTime" onClick={this.reset}>
              Descartar Tempo
            </button>
          </div>
        </div>
      </div>
    );
  }
}

export default Stopwatch;
