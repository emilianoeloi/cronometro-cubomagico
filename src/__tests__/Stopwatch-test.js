import React from 'react';
import renderer from 'react-test-renderer';

import Stopwatch from '../Stopwatch';

test('test table of Times', () => {
  const component = renderer.create(
    <Stopwatch />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
