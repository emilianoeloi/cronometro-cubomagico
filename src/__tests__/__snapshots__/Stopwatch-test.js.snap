exports[`test test table of Times 1`] = `
<div
  className="pure-g">
  <div
    className="pure-u-1">
    <h2
      id="stopwatch-display">
      00\' 00.00\"
    </h2>
  </div>
  <div
    className="pure-u-1">
    <div
      className="l-box">
      <button
        className="pure-button button-start"
        id="stopwatch-start"
        onClick={[Function]}>
        Iniciar
      </button>
       
      <button
        className="pure-button button-stop"
        id="stopwatch-stop"
        onClick={[Function]}>
        Parar
      </button>
    </div>
  </div>
  <div
    className="pure-u-1">
    <div
      className="l-box">
      <button
        className="pure-button pure-button-active"
        disabled={true}
        id="stopwatch-saveTime"
        onClick={[Function]}>
        Salvar Tempo
      </button>
       
      <button
        className="pure-button pure-button-active"
        id="stopwatch-discardTime"
        onClick={[Function]}>
        Descartar Tempo
      </button>
    </div>
  </div>
</div>
`;
